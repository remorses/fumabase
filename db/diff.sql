-- AlterTable
ALTER TABLE "MarkdownPage" ALTER COLUMN "githubSha" DROP NOT NULL;

-- AlterTable
ALTER TABLE "SiteBranch" DROP COLUMN "trieveDatasetId",
DROP COLUMN "trieveReadApiKey";

