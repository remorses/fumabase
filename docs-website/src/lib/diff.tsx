/* diffHighlight.ts ---------------------------------------------------- */
import { visit } from 'unist-util-visit'
import { diff_match_patch, DIFF_INSERT } from 'diff-match-patch'
import type { Root, Parent, Literal, Text, Content as MdContent } from 'mdast'

declare module 'mdast' {
    export interface Data {
        hProperties?: {
            id?: string
            'data-added'?: boolean | string
        }
    }
}

/* -------------------------------------------------------------------- *
 * 1.  MARK ADDITIONS (adds data-added attr to new or edited nodes)
 * -------------------------------------------------------------------- */

export function markRemarkAstAdditions(oldTree: Root, newTree: Root): void {
    /* pass 1 - fingerprint every node that already existed */
    const previous = new Set<string>()
    visit(oldTree, (n) => {
        previous.add(fp(n))
    })

    /* pass 2 - walk new tree, tag additions + diff text */
    visit(newTree, (node: MdContent | Parent) => {
        if (!previous.has(fp(node))) setAdded(node)
        else if (node.type === 'text') diffText(node as Text, oldTree)
    })
}

/* ---------- helpers ------------------------------------------------- */

// stable  fingerprint: ignore layout data + actual text content
function fp(n: MdContent | Parent): string {
    const { type } = n
    const base: Record<string, unknown> = { type }

    switch (type) {
        case 'text':
            /* same position = same identity for a text node */
            base.pos = n.position?.start?.offset
            break
        case 'link':
            base.url = n['url']
            break
        default:
            break
    }
    return JSON.stringify(base)
}

function setAdded(n: MdContent | Parent): void {
    ;(n.data ??= {}).hProperties ??= {}
    ;(n.data.hProperties as Record<string, unknown>)['data-added'] = true
}

function diffText(newText: Text, oldTree: Root): void {
    const oldNode = findSameSpotText(newText, oldTree)
    if (!oldNode || oldNode.value === newText.value) return

    const dmp = new diff_match_patch()
    const diffs = dmp.diff_main(oldNode.value, newText.value)
    dmp.diff_cleanupSemantic(diffs)

    const pieces: Text[] = diffs.flatMap(([op, data]): Text[] => {
        if (op === DIFF_INSERT) {
            return [
                {
                    type: 'text',
                    value: data,
                    data: { hProperties: { 'data-added': true } },
                },
            ]
        }
        if (op === 0) return [{ type: 'text', value: data }]
        return []
    })

    // swap text node for an inline paragraph of pieces
    Object.assign(newText, {
        type: 'paragraph',
        children: pieces,
        value: undefined,
    })
}

function findSameSpotText(target: Text, tree: Root): Text | null {
    let result: Text | null = null
    visit(tree, (n: MdContent | Parent) => {
        if (
            !result &&
            n.type === 'text' &&
            n.position?.start?.offset === target.position?.start?.offset
        ) {
            result = n as Text
        }
    })
    return result
}

/* -------------------------------------------------------------------- *
 * 2.  REACT HOOK â‡’ APPLY CSS CUSTOM HIGHLIGHTS
 * -------------------------------------------------------------------- */

import { useLayoutEffect, useRef, RefObject } from 'react'
export const useAddedHighlighter = ({
    root,
    enabled = true,
}: {
    root: RefObject<HTMLElement | null>
    enabled?: boolean
}): void => {
    const hlRef = useRef<Highlight | null>(null)

    useLayoutEffect(() => {
        if (!enabled) return

        const host = root.current
        if (!host || !CSS.highlights) return

        /* collect all added nodes except opt-outs */
        const added: HTMLElement[] = Array.from(
            host.querySelectorAll<HTMLElement>(
                '[data-added]:not([data-no-highlight])',
            ),
        )

        if (!added.length) return

        const ranges = added.map((el) => {
            const r = document.createRange()
            r.selectNodeContents(el)
            return r
        })

        const hl = hlRef.current ?? new window.Highlight()
        ranges.forEach((r) => hl.add(r))
        CSS.highlights.set('md-add', hl)
        hlRef.current = hl

        return () => {
            /* optional cleanup on unmount */
            CSS.highlights.delete('md-add')
            hlRef.current = null
        }
    })
}
