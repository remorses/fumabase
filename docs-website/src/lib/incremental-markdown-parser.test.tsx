import { toMarkdown } from 'mdast-util-to-markdown'
import { describe, expect, test } from 'vitest'

import { diffWords } from 'diff'
import {
    IncrementalParsingProps,
    parseMarkdownIncremental,
    SegmentEntry,
} from './incremental-markdown-parser'

// Wait for highlighter to initialize
async function parseAndWaitForHighlighter(props: IncrementalParsingProps) {
    let retries = 0
    while (retries < 10) {
        try {
            return parseMarkdownIncremental({
                ...props,
            })
        } catch (promise) {
            if (promise instanceof Promise) {
                await promise
                retries++
            } else {
                throw promise
            }
        }
    }
    throw new Error('Highlighter initialization timed out')
}

function simulateStreaming(text: string): string[] {
    const words = text.split(' ')
    const result: string[] = []
    for (let i = 1; i <= words.length; i++) {
        result.push(words.slice(0, i).join(' '))
    }
    return result
}

describe('getOptimizedMarkdownAst', () => {
    const cache = new Map<number, SegmentEntry>()
    const extension = 'mdx'
    const markdown = `
# Welcome to the Example Document

This is a sample markdown file to showcase various features.

## Features

* Easy formatting
* Lists
* **Bold text**
* [Links](https://example.com)

> "Markdown makes documentation easy!"

\`\`\`js
console.log('Hello, world!');
\`\`\`

1. one
2. two
3. three

this is another paragraph


`.trim()

    // Loop over streaming chunks and parse incrementally
    const streamingChunks = simulateStreaming(markdown)
    streamingChunks.forEach((chunk, idx) => {
        test(`incremental parsing at chunk ${idx + 1}`, async () => {
            const ast = await parseAndWaitForHighlighter({
                markdown: chunk,
                cache,
                extension,
            })

            const md = toMarkdown(ast)
            // const changes = diffWords(chunk, md)
            // const additions = changes.filter((change) => change.added)
            expect(md).toMatchSnapshot()
        })
    })

    test('returns full AST when no previous data', async () => {
        const ast = await parseAndWaitForHighlighter({
            markdown,
            cache,
            extension,
        })

        expect(toMarkdown(ast)).toMatchInlineSnapshot(`
          "This is a sample markdown file to showcase various features.

          ## Features

          * Easy formatting
          * Lists
          * **Bold text**
          * [Links](https://example.com)

          > "Markdown makes documentation easy!"

          \`\`\`js
          console.log('Hello, world!');
          \`\`\`

          1. one
          2. two
          3. three

          this is another paragraph
          "
        `)
    })
})
