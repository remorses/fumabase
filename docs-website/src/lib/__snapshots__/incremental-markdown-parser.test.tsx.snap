// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getOptimizedMarkdownAst > incremental parsing at chunk 1 1`] = `""`;

exports[`getOptimizedMarkdownAst > incremental parsing at chunk 2 1`] = `""`;

exports[`getOptimizedMarkdownAst > incremental parsing at chunk 3 1`] = `""`;

exports[`getOptimizedMarkdownAst > incremental parsing at chunk 4 1`] = `""`;

exports[`getOptimizedMarkdownAst > incremental parsing at chunk 5 1`] = `""`;

exports[`getOptimizedMarkdownAst > incremental parsing at chunk 6 1`] = `
"This
"
`;

exports[`getOptimizedMarkdownAst > incremental parsing at chunk 7 1`] = `
"This is
"
`;

exports[`getOptimizedMarkdownAst > incremental parsing at chunk 8 1`] = `
"This is a
"
`;

exports[`getOptimizedMarkdownAst > incremental parsing at chunk 9 1`] = `
"This is a sample
"
`;

exports[`getOptimizedMarkdownAst > incremental parsing at chunk 10 1`] = `
"This is a sample markdown
"
`;

exports[`getOptimizedMarkdownAst > incremental parsing at chunk 11 1`] = `
"This is a sample markdown file
"
`;

exports[`getOptimizedMarkdownAst > incremental parsing at chunk 12 1`] = `
"This is a sample markdown file to
"
`;

exports[`getOptimizedMarkdownAst > incremental parsing at chunk 13 1`] = `
"This is a sample markdown file to showcase
"
`;

exports[`getOptimizedMarkdownAst > incremental parsing at chunk 14 1`] = `
"This is a sample markdown file to showcase various
"
`;

exports[`getOptimizedMarkdownAst > incremental parsing at chunk 15 1`] = `
"This is a sample markdown file to showcase various features.

##
"
`;

exports[`getOptimizedMarkdownAst > incremental parsing at chunk 16 1`] = `
"This is a sample markdown file to showcase various features.

## Features

*
"
`;

exports[`getOptimizedMarkdownAst > incremental parsing at chunk 17 1`] = `
"This is a sample markdown file to showcase various features.

## Features

* Easy
"
`;

exports[`getOptimizedMarkdownAst > incremental parsing at chunk 18 1`] = `
"This is a sample markdown file to showcase various features.

## Features

* Easy formatting
*
"
`;

exports[`getOptimizedMarkdownAst > incremental parsing at chunk 19 1`] = `
"This is a sample markdown file to showcase various features.

## Features

* Easy formatting
* Lists
*
"
`;

exports[`getOptimizedMarkdownAst > incremental parsing at chunk 20 1`] = `
"This is a sample markdown file to showcase various features.

## Features

* Easy formatting
* Lists
* \\*\\*Bold
"
`;

exports[`getOptimizedMarkdownAst > incremental parsing at chunk 21 1`] = `
"This is a sample markdown file to showcase various features.

## Features

* Easy formatting
* Lists
* **Bold text**
*
"
`;

exports[`getOptimizedMarkdownAst > incremental parsing at chunk 22 1`] = `
"This is a sample markdown file to showcase various features.

## Features

* Easy formatting
* Lists
* **Bold text**
* [Links](https://example.com)

>
"
`;

exports[`getOptimizedMarkdownAst > incremental parsing at chunk 23 1`] = `
"This is a sample markdown file to showcase various features.

## Features

* Easy formatting
* Lists
* **Bold text**
* [Links](https://example.com)

> "Markdown
"
`;

exports[`getOptimizedMarkdownAst > incremental parsing at chunk 24 1`] = `
"This is a sample markdown file to showcase various features.

## Features

* Easy formatting
* Lists
* **Bold text**
* [Links](https://example.com)

> "Markdown makes
"
`;

exports[`getOptimizedMarkdownAst > incremental parsing at chunk 25 1`] = `
"This is a sample markdown file to showcase various features.

## Features

* Easy formatting
* Lists
* **Bold text**
* [Links](https://example.com)

> "Markdown makes documentation
"
`;

exports[`getOptimizedMarkdownAst > incremental parsing at chunk 26 1`] = `
"This is a sample markdown file to showcase various features.

## Features

* Easy formatting
* Lists
* **Bold text**
* [Links](https://example.com)

> "Markdown makes documentation easy!"

\`\`\`js
console.log('Hello,
\`\`\`
"
`;

exports[`getOptimizedMarkdownAst > incremental parsing at chunk 27 1`] = `
"This is a sample markdown file to showcase various features.

## Features

* Easy formatting
* Lists
* **Bold text**
* [Links](https://example.com)

> "Markdown makes documentation easy!"

\`\`\`js
console.log('Hello, world!');
\`\`\`

1.
"
`;

exports[`getOptimizedMarkdownAst > incremental parsing at chunk 28 1`] = `
"This is a sample markdown file to showcase various features.

## Features

* Easy formatting
* Lists
* **Bold text**
* [Links](https://example.com)

> "Markdown makes documentation easy!"

\`\`\`js
console.log('Hello, world!');
\`\`\`

1. one
2.
"
`;

exports[`getOptimizedMarkdownAst > incremental parsing at chunk 29 1`] = `
"This is a sample markdown file to showcase various features.

## Features

* Easy formatting
* Lists
* **Bold text**
* [Links](https://example.com)

> "Markdown makes documentation easy!"

\`\`\`js
console.log('Hello, world!');
\`\`\`

1. one
2. two
3.
"
`;

exports[`getOptimizedMarkdownAst > incremental parsing at chunk 30 1`] = `
"This is a sample markdown file to showcase various features.

## Features

* Easy formatting
* Lists
* **Bold text**
* [Links](https://example.com)

> "Markdown makes documentation easy!"

\`\`\`js
console.log('Hello, world!');
\`\`\`

1. one
2. two
3. three

this
"
`;

exports[`getOptimizedMarkdownAst > incremental parsing at chunk 31 1`] = `
"This is a sample markdown file to showcase various features.

## Features

* Easy formatting
* Lists
* **Bold text**
* [Links](https://example.com)

> "Markdown makes documentation easy!"

\`\`\`js
console.log('Hello, world!');
\`\`\`

1. one
2. two
3. three

this is
"
`;

exports[`getOptimizedMarkdownAst > incremental parsing at chunk 32 1`] = `
"This is a sample markdown file to showcase various features.

## Features

* Easy formatting
* Lists
* **Bold text**
* [Links](https://example.com)

> "Markdown makes documentation easy!"

\`\`\`js
console.log('Hello, world!');
\`\`\`

1. one
2. two
3. three

this is another
"
`;

exports[`getOptimizedMarkdownAst > incremental parsing at chunk 33 1`] = `
"This is a sample markdown file to showcase various features.

## Features

* Easy formatting
* Lists
* **Bold text**
* [Links](https://example.com)

> "Markdown makes documentation easy!"

\`\`\`js
console.log('Hello, world!');
\`\`\`

1. one
2. two
3. three

this is another paragraph
"
`;
